File: TypeScript/app/gilded-rose.ts
export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name, sellIn, quality) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  updateQuality() {
    for (let i = 0; i < this.items.length; i++) {
      let decrement = 1;
      if (this.items[i].name.includes("Conjured")) {
        decrement = 2;
      }
      switch (this.items[i].name) {
        case 'Aged Brie':
          this.incrementQuality(this.items[i]);
          break;
        case 'Backstage passes to a TAFKAL80ETC concert':
          this.updateBackstageQuality(this.items[i]);
          break;
        case 'Sulfuras, Hand of Ragnaros':
          break;
        default:
          this.decrementQuality(this.items[i], decrement);
          break;
      }

      if (this.items[i].name !== 'Sulfuras, Hand of Ragnaros') {
        this.items[i].sellIn--;
      }

      if (this.items[i].sellIn < 0) {
        switch (this.items[i].name) {
          case 'Aged Brie':
            this.incrementQuality(this.items[i]);
            break;
          case 'Backstage passes to a TAFKAL80ETC concert':
            this.items[i].quality = 0;
            break;
          case "Sulfuras, Hand of Ragnaros":
            break;
          default:
            this.decrementQuality(this.items[i], decrement);
            break;
        }
      }
    }

    return this.items;
  }

  private decrementQuality(item: Item, factor:number = 1) {
    if (item.quality > 0) {
      item.quality-= factor;
    }
  }

  private incrementQuality(item: Item) {
    if (item.quality < 50) {
      item.quality++;
    }
  }

  private updateBackstageQuality(item: Item) {
    this.incrementQuality(item);
    if (item.sellIn < 11) {
      this.incrementQuality(item);
    }
    if (item.sellIn < 6) {
      this.incrementQuality(item);
    }
  }
}

File: python/gilded_rose.py
# -*- coding: utf-8 -*-

class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality


class GildedRose:
    def __init__(self, items=[]):
        self.items = items

    def update_quality(self):
        for item in self.items:
            if item.name != "Aged Brie" and item.name != "Backstage passes to a TAFKAL80ETC concert":
                factor = 1
                if "Conjured" in item.name:
                    factor = 2
                self.decrement_quality(item, factor)
            else:
                self.increment_quality(item)
                if item.name == "Backstage passes to a TAFKAL80ETC concert":
                    self.update_backstage_quality(item)
            if item.name != "Sulfuras, Hand of Ragnaros":
                item.sell_in = item.sell_in - 1
            print(item.name + ", " + str(item.sell_in) + ", " + str(item.quality))

    @staticmethod
    def decrement_quality(item, factor=1):
        if item.quality > 0:
            item.quality -= factor

    @staticmethod
    def increment_quality(item):
        if item.quality < 50:
            item.quality += 1

    def update_backstage_quality(self, item):
        self.increment_quality(item)
        if item.sell_in < 11:
            self.increment_quality(item)
        if item.sell_in < 6:
            self.increment_quality(item)
